### PR TITLE
Filter calculations animation

### PR DESCRIPTION
**Pull request by TheDarkLord247**
This component includes an animation that shows how filters are calculated.

I've used the[ react-slick](https://react-slick.neostack.com/) package to help with the animation.
Installed the necessary dependencies and made small background color changes to Maya's FilterSizes component

### CODE REVIEW

**Comment by TheDarkLord247** [comment]
Working on a small change to the buttons - gonna push the change in a bit

File: src/components/Carousel.js
Comment (krashanoff) [inline src/components/Carousel.js:45]: [Alttext](https://webaim.org/techniques/alttext/) here should be a little more descriptive and concise, and describe what each step is briefly.

File: src/components/Carousel.js
Comment (krashanoff) [inline src/components/Carousel.js:43]: This is the part that irked me: `this.slider = slider`

**Review by krashanoff — CHANGES_REQUESTED** [review]
Hey Aman! This is looking good. Just gave it a quick once-over. Big piece of commentary here is that I think that the `slick-carousel` component is a little dicey. Just looking at how it renders out:

source of rendered slick carousel

It ignores your styling as far as I can tell.

It really is only playing with the slides by transitioning their opacity in and out on a loop around its children. The dots could be rendered with some FontAwesome icons or empty buttons, and the state could preserve the current slide information, etc. Here's how I would go about writing a quick one:
* Create a Carousel component.
* For each child of `props.children` (the components drawn within the component's tag), iterate over it, setting opacity of all to zero.
* Set an interval, iterating over the children and applying some class to the current one that makes its opacity 1.

If we play our cards right, this could prevent you from writing some unnecessary CSS, keep you in control of the flow, and keep things more in-line with React.

If you'd like a code example, here's a basic one.

```js
// Carousel.js
import React, { useEffect, useState } from 'react';
import './Carousel.css';

export default function Carousel(props) {
  const [ currentSlide, setCurrentSlide ] = useState(0);

  useEffect(() => {
    const interval = setInterval(() => {
      setCurrentSlide((currentSlide + 1) % props.children.length);
    }, 1000);

    return () => clearInterval(interval);
  });

  return (
    <div className='carousel'>
      {props.children.map((c, i) => <div key={i} className={`carousel-slide ${i === currentSlide ? 'visible' : ''}`}>{c}</div>)}
    </div>
  );
}
```

And for the styling:

```css
/* Carousel.css */
.carousel > * {
    position: absolute;
}

.carousel-slide {
    opacity: 0;
    transition: opacity .3s ease;
}

.visible {
    opacity: 1;
}
```

File: src/components/Carousel.js
Comment (TheDarkLord247) [reply src/components/Carousel.js:43]: What is generally the problem with inline styling? Does it take too long to render for older browsers/poorer connections? 

And if I understand correctly, the alternate route is making a carousel, setting all the pictures' opacity to 0, and then making each picture's opacity 1 after a particular interval?

Also, one final question about the syntax: I don't understand the props.children.map => <div> syntax here 
```
        {props.children.map((c, i) => <div key={i} className={`carousel-slide ${i === currentSlide ? 'visible' : ''}`}>{c}</div>)}
```

File: src/components/Carousel.js
Comment (TheDarkLord247) [reply src/components/Carousel.js:45]: Since all the frames are showing basically the same thing, is it okay to give a complete description for the first image and then just say second example, third example, etc. for the following pictures?

File: src/components/Carousel.js
Comment (krashanoff) [reply src/components/Carousel.js:43]: > What is generally the problem with inline styling? Does it take too long to render for older browsers/poorer connections?

There actually isn't much a problem with inline styling! I only pointed it out because it is something that is indicative of some dicey code. Inline styles override any existing ones in your stylesheets, and often lead to unpredictable problems between components.

> And if I understand correctly, the alternate route is making a carousel, setting all the pictures' opacity to 0, and then making each picture's opacity 1 after a particular interval?

Yes, the way that the example carousel works is by setting the opacity of every picture to zero by default (gives every child element the `.carousel-slide` class), then applies the `.visible` class to the "current slide", which is determined by the state variable `currentSlide`. We then leverage `useEffect` to set an interval that increments that `currentSlide` state variable by one every handful of milliseconds, wrapping around to zero when we exceed the number of children passed to the component.

The kicker here is that we draw all the elements on top of each other by adding a container class that positions all child elements (`.carousel > *`) absolutely in the same location.

> I don't understand the props.children.map =>
> syntax here

This is a pretty dense line, so I'll break it up into a more verbose set and comment each one. Let me know if there's anything that's unclear!

```js
import React, { useEffect, useState } from 'react';
import './Example.css';

export default function Carousel(props) {
  const [ currentSlide, setCurrentSlide ] = useState(0);

  useEffect(() => {
    const interval = setInterval(() => {
      setCurrentSlide((currentSlide + 1) % props.children.length);
    }, 1000);

    return () => clearInterval(interval);
  });

  // Every React component has a property passed down to it silently:
  // props.children.
  // This is an array of all the **child** components passed to it.
  // e.g.:
  //
  // <Carousel>
  //   <h1>First child</h1>
  //   <h2>Second child</h2>
  // </Carousel>
  //
  // Here, our children are the <h1/> and <h2/> tags.

  // Let's create an Array to hold our mutated children.
  let newChildren = [];

  // Now, we loop through props.children and mutate each
  // component.
  for (let i = 0; i < props.children.length; i++) {
    // Our current child element will be 'c'.
    const c = props.children[i];

    // We need to create a class string. Each component
    // needs to have class 'carousel-slide', though, so
    // we start from there.
    let classString = 'carousel-slide';

    // If our current child that we're iterating through
    // is the current slide, it should have the 'visible'
    // class. Append it with a leading space (to separate
    // class names).
    if (i === currentSlide)
      classString += ' visible';

    // Then, construct our child and push it onto our list.
    newChildren.push(
      // Each child needs to be wrapped with another div with
      // our provided classes.
      // Since it is a member of an array, it also needs a unique
      // 'key' prop. We can just use the index of the item.
      <div key={i} className={classString}>
        {c}
      </div>
    )
  }
  
  return (
    <div className='carousel'>
      {newChildren}
    </div>
  );
}
```

File: src/components/Carousel.js
Comment (krashanoff) [reply src/components/Carousel.js:45]: The idea is that you are trying to explain the process step by step through these images, right? The alttext should describe which step each image is conveying to the user. To better visualize this, here's a look at the carousel without images:

carousel of alttext

Your idea of describing the first image, then using something like "computing element 2,1" for the following alttext is a good idea. You don't need to be too verbose though, just a *brief* description of the operation is more than plenty.

Recall that a critical feature of alttext is that it is contextual. If you have already explained what "multiplying each element of a matrix" is at length, then you shouldn't need to do it again. Hope this answered your question!

File: src/components/Carousel.js
Comment (TheDarkLord247) [reply src/components/Carousel.js:43]: I understand now, thank you so much!

Comment (TheDarkLord247) [reply src/components/Carousel.js:43]: Is it alright to stick to react-slick for now and change it later if there are any issues? Because I'm currently working on making new frames (the current ones are slightly inconsistent)

**Comment by TheDarkLord247** [comment]
Getting this weird compilation error:
"Redundant alt attribute. Screen-readers already announce `img` tags as an image. You don’t need to use the words `image`, `photo,` or `picture` (or any specified custom words) in the alt prop  jsx-a11y/img-redundant-alt"

My alt message is "first example of feature pool calculation - shows how the first element of the feature map is calculated by taking the dot product of the filter and a portion of our image" and I need the word image to be in the alt message.

**Comment by malsf21** [comment]
Oh yeah, I'm not a huge fan of how that is set up. You can disable that with `eslint`'s disable features; [see the ESLint docs](https://eslint.org/docs/2.13.1/user-guide/configuring#disabling-rules-with-inline-comments-1) for more information.

File: src/components/Carousel.js
Comment (krashanoff) [reply src/components/Carousel.js:43]: If you'd like to stick with `react-slick`, that's alright, but I would discourage it.

The main point of my example here was to point out that making your own isn't as gnarly as it sounds, and offers more control over the process without pulling some alarming cards.

Comment (TheDarkLord247) [reply src/components/Carousel.js:43]: I can definitely work on the carousel alongside but don't know if I'll have the time for it since I'm working on new frames for the carousel. I'll try my best to switch away from react-slick

**Comment by TheDarkLord247** [comment]
Just to summarize:
- I understand that using `react-slick` is not ideal and will try to make my own carousel in the next week or so to replace it.
- I used ESLint's no-alert method to get around the alert message I was getting for my images' alt messages.

File: src/components/Carousel.js
Comment (malsf21) [inline src/components/Carousel.js:2]: I don't think this is the right syntax (if you notice, it executes the alert) - you should instead do something like

```js
/* eslint-disable jsx-a11y/img-redundant-alt */
```

**Review by malsf21 — CHANGES_REQUESTED** [review]

**Comment by malsf21** [comment]
Also we may want to discuss the visual presentation of the carousel - we may do this at the PM meeting or in some other form. Don't think that's a blocker to merge right now.

File: src/components/Carousel.js
Comment (TheDarkLord247) [reply src/components/Carousel.js:2]: Oh okay, It didnt give me an error so I thought I was good XD. Thanks for that!

**Review by malsf21 — APPROVED** [review]
I'll defer the rest to Leo on whether or not we should merge now, but for now I think we're at a breakpoint where we can add stuff. Slick refactor may be nice though, I'm never a fan of carousel implementations 😟

**Comment by TheDarkLord247** [comment]
> Also we may want to discuss the visual presentation of the carousel - we may do this at the PM meeting or in some other form. Don't think that's a blocker to merge right now.
What do you mean by this?

**Review by krashanoff — APPROVED** [review]
> Also we may want to discuss the visual presentation of the carousel - we may do this at the PM meeting or in some other form. Don't think that's a blocker to merge right now.

I think Matt just meant that it might be worth investigating the best way of presenting the slides to get the process across. For example, right now, the slides just scroll through one by one, but it may be beneficial to lay out the slides side-by-side with some blurbs about the process at each step. We want to make sure that these learning labs help make the material intuitive :+1:

> I'll defer the rest to Leo on whether or not we should merge now, but for now I think we're at a breakpoint where we can merge. Slick refactor may be nice though, I'm never a fan of carousel implementations 😟

Agreed. This is a good breakpoint since we will roll with Slick for the time being. If we run into problems down the line, a refactor may be in order, but for now this looks good. Thanks Aman!

**Comment by malsf21** [comment]
Ok sounds good! Aman you're good to merge, squash merge if you can 😄

**Comment by TheDarkLord247** [comment]
> I think Matt just meant that it might be worth investigating the best way of presenting the slides to get the process across. For example, right now, the slides just scroll through one by one, but it may be beneficial to lay out the slides side-by-side with some blurbs about the process at each step. We want to make sure that these learning labs help make the material intuitive 👍

We go through the process of calculation in-depth in our slides (taught in session 4b), and our idea was to have people first attend that session/go through the slides and then go through the Learning Lab!

**Comment by TheDarkLord247** [comment]
Thanks for all your help and efforts! :D @malsf21 @krashanoff

**Pull request merged**
